index.js

```diff
@@ -16,25 +16,51 @@ export default function nanoSpawn(file, commandArguments = [], options = {}) {
 	const command = getCommand(file, commandArguments);
 	const spawnOptions = getOptions(options);
 	[file, commandArguments] = handleNode(file, commandArguments);
-	const forcedShell = getForcedShell(file, spawnOptions);
-	spawnOptions.shell ||= forcedShell;
 	const input = getInput(spawnOptions);
 
-	const subprocess = spawn(...escapeArguments(file, commandArguments, forcedShell), spawnOptions);
-
-	useInput(subprocess, input);
-	const resultPromise = getResult(subprocess, start, command);
+	const instancePromise = getInstance({
+		file,
+		commandArguments,
+		spawnOptions,
+		start,
+		command,
+	});
+	const resultPromise = getResult(instancePromise, input, start, command);
+	const stdoutLines = lineIterator(instancePromise, 'stdout', resultPromise);
+	const stderrLines = lineIterator(instancePromise, 'stderr', resultPromise);
 
-	const stdoutLines = lineIterator(subprocess.stdout, resultPromise);
-	const stderrLines = lineIterator(subprocess.stderr, resultPromise);
 	return Object.assign(resultPromise, {
-		subprocess,
+		nodeChildProcess: instancePromise,
 		[Symbol.asyncIterator]: () => combineAsyncIterators(stdoutLines, stderrLines),
 		stdout: stdoutLines,
 		stderr: stderrLines,
 	});
 }
 
+const getInstance = async ({file, commandArguments, spawnOptions, start, command}) => {
+	try {
+		const forcedShell = await getForcedShell(file, spawnOptions);
+		spawnOptions.shell ||= forcedShell;
+		const instance = spawn(...escapeArguments(file, commandArguments, forcedShell), spawnOptions);
+
+		// The `error` event is caught by `once(instance, 'spawn')` and `once(instance, 'close')`.
+		// But it creates an uncaught exception if it happens exactly one tick after 'spawn'.
+		// This prevents that.
+		instance.once('error', () => {});
+
+		await once(instance, 'spawn');
+		return instance;
+	} catch (error) {
+		throw getResultError({
+			error,
+			result: initResult(),
```

Comment:
If the subprocess throws during `spawn()`, we don't have access to the `.stdout` and `.stderr` stream properties. So the only way to know whether `result.stdout`/`result.stderr` should be `undefined` or not is to check whether `spawnOptions.stdout` is either `"pipe"` or `"overlapped"`.

We could do that check, but to save some bytes, this PR is making `result.stdout`/`result.stderr` an empty string instead of `undefined` when using things like `stdout: 'ignore'` or `stdout: 'inherit'`. Also, this could make things a little easier with types. Finally, it could be slightly simpler for users to assume those properties are always strings.

On the other hand, there are some benefits in the `undefined` vs empty string distinction that Execa is making. So I'm not sure whether that's a good idea or not, please let me know what you think.

Reply:
I think empty strings are the best choice here, and would be for `execa` too.